sdk/storage/AzureStorageBlob/Source/DataStructures/StorageStreamUploader.swift

```diff
@@ -0,0 +1,540 @@
+        guard !isComplete else { return }
+        let metadata = blockList.removeFirst()
+        let range = metadata.range
+        let blockId = metadata.blockId
+        let uploader = ChunkUploader(
+            blob: blobName,
+            container: containerName,
+            blockId: blockId,
+            client: client,
+            url: uploadSource,
+            startRange: range.startIndex,
+            endRange: range.endIndex,
+            options: options
+        )
+        uploader.upload { result, httpResponse in
+            switch result {
+            case .success:
+                // Add block ID to the completed list and lookup where its final
+                // placement should be
+                self.progress += uploader.endRange - uploader.startRange
+                let blockId = uploader.blockId
+                self.completedBlockMap[blockId] = self.blockIdMap[blockId]
+            case let .failure(error):
+                self.client.options.logger.debug(String(describing: error))
+            }
+            completion(result, httpResponse)
+            group?.leave()
+        }
+    }
+
+    // MARK: Private Methods
+
+    private func buildLookupList() -> BlobLookupList {
+        let sortedIds = completedBlockMap.sorted(by: { $0.value < $1.value })
+        let lookupList = BlobLookupList(latest: sortedIds.compactMap { $0.key })
+        return lookupList
+    }
+
+    private func computeBlockList(withOffset offset: Int = 0) -> [(Range<Int>, String)] {
+        var blockList = [(Range<Int>, String)]()
+        let alignForCrypto = isEncrypted
+        let chunkLength = client.options.maxChunkSize
+        let start = offset
+        guard let end = fileSize else {
+            fatalError("File size not found.")
+        }
+
+        if alignForCrypto {
+            fatalError("Client-side encryption is not yet supported!")
+        } else {
+            for index in stride(from: start, to: end, by: chunkLength) {
+                var end = index + chunkLength
+                if let fileSize = fileSize, end > fileSize {
+                    end = fileSize
+                }
+                let range = index ..< end
+                let blockId = UUID().uuidString.base64String
+                blockList.append((range: range, blockId: blockId))
+                blockIdMap[blockId] = blockList.count - 1
+            }
+        }
+        return blockList
+    }
+
+    /// Parses the blob length from the content range header: bytes 1-3/65537
+    private func parseLength(fromContentRange contentRange: String?) throws -> Int {
+        let error = AzureError.general("Unable to parse content range: \(contentRange ?? "nil")")
+        guard let contentRange = contentRange else { throw error }
+        // First, split in space and take the second half: "1-3/65537"
+        guard let byteString = contentRange.split(separator: " ", maxSplits: 1).last else { throw error }
```

Comment:
Arguably this step isn't needed, but I don't feel strongly about removing it.

Reply:
Good point! Fixed.